### PR TITLE
Added support for continuation frames

### LINKED ISSUE #1350: websockets does not support continuation frames
The uwsgi_websocket_recv_do() function treats opcode %x0 the same as opcode %x1 or %x2. This results in continuation frames getting to the application as if they are separate websockets messages. Because the application can't know what the opcode was, it can't aggregate frames into complete messages itself.

This results in a situation where large JSONs are received in pieces and cannot be parsed.

I'm not a C developer, so I might be mistaken, but I couldn't find any check for the FIN flag in the receive function as well.

To verify the issue, I recompiled uWSGI with a log message for the received opcode, just before running parse:

```
uwsgi_log("[uwsgi-websocket] \"%.*s %.*s\" (%.*s) received opcode: %d \n", REQ_DATA, wsgi_req->websocket_opcode );
return uwsgi_websockets_parse(wsgi_req);
```

I then ran the websockets_chat_2 test, and pasted Descartes' meditations from Project Gutenberg:
http://www.gutenberg.org/cache/epub/23306/pg23306.txt
In the middle of page 113, the first frame ends, and a continuation message is received with opcode %x0. That second frame is sent to the websockets_chat_2 application as a separate message.

### PR DESCRIPTION
**Pull request by timitoc**
This should resolve #1350

### CODE REVIEW

File: core/websockets.c [hunk 1]
```diff
@@ -161,14 +162,38 @@ static struct uwsgi_buffer *uwsgi_websockets_parse(struct wsgi_request *wsgi_req
...
+
+	if (wsgi_req->websocket_is_fin) {
+		uwsgi.websockets_continuation_buffer = NULL;
+		/// Freeing websockets_continuation_buffer is done by the caller
+		return ub;
+	}
+	uwsgi.websockets_continuation_buffer = ub;
+	/// Message is not complete, send empty dummy buffer to signal waiting for full message
+	ub = uwsgi_buffer_new(1);
+        uwsgi_buffer_append(ub, "\0", 1);
```
Comment (xrmx) [inline core/websockets.c]: tabs please

File: uwsgi.h [hunk 2]
```diff
@@ -2765,6 +2766,7 @@ struct uwsgi_server {
 
 	struct uwsgi_buffer *websockets_ping;
 	struct uwsgi_buffer *websockets_pong;
+	struct uwsgi_buffer *websockets_continuation_buffer;
```
Comment (xrmx) [inline uwsgi.h]: new members at the end of the struct

File: uwsgi.h [hunk 3]
```diff
@@ -1569,6 +1569,7 @@ struct wsgi_request {
 	size_t websocket_need;
 	int websocket_phase;
 	uint8_t websocket_opcode;
+	uint8_t websocket_is_fin;
```
Comment (xrmx) [inline uwsgi.h]: same as below

**Comment by ecamaj** [comment]
@xrmx Please merge this to branch 2-0

**Comment by ecamaj** [comment]
@unbit ping.

**Comment by ecamaj** [comment]
@rdeioris ping

**Comment by suhjohn** [comment]
@ecamaj I also had concerns regarding this issue and it'd be great to see this get merged. May I ask how you tested the code? Did you force continuation message from a testing client, or did you create unit tests?

**Comment by ecamaj** [comment]
Yeah, I have client app that uploads JSON with base64 image in it. I’m using this fix in production since pull request is created, works like a charm.

**Comment by ewoudwerkman** [comment]
Any progress on merging this into a release?

I'm currently using long polling to circumvent this issue, but that doesn't scale and perform well...

**Comment by jpboily-axceta** [comment]
We had an issue when trying to send huge payload to our backend through websocket. The payload was truncated by uWSGI. I implemented this PR in a custom branch and it solved our problem.

**Comment by ewoudwerkman** [comment]
@unbit All checks have passed, can you merge this pull request?

**Comment by rdeioris** [comment]
@xrmx safe for backport but pay attention to the new uwsgi_server field

**Comment by ewoudwerkman** [comment]
@rdeioris Thanks!
What are the plans for a next point release?

**Comment by ecamaj** [comment]
That would be really nice, to release new version...

**Pull request merged**